### Задание 1


Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.
 

In [2]:
import pandas as pd
import numpy as np

In [49]:
horse = pd.read_csv(f'd:\workspace\horse_data.csv', names=[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28], na_values = '?')

In [50]:
df1 = horse.iloc[:,3:11]

In [9]:
df1.describe()

4           5           6           7           8           9   \
count  240.000000  276.000000  242.000000  244.000000  231.000000  253.000000   
mean    38.167917   71.913043   30.417355    2.348361    2.017316    2.853755   
std      0.732289   28.630557   17.642231    1.045054    1.042428    1.620294   
min     35.400000   30.000000    8.000000    1.000000    1.000000    1.000000   
25%     37.800000   48.000000   18.500000    1.000000    1.000000    1.000000   
50%     38.200000   64.000000   24.500000    3.000000    2.000000    3.000000   
75%     38.500000   88.000000   36.000000    3.000000    3.000000    4.000000   
max     40.800000  184.000000   96.000000    4.000000    4.000000    6.000000   

               10         11  
count  268.000000  245.00000  
mean     1.305970    2.95102  
std      0.477629    1.30794  
min      1.000000    1.00000  
25%      1.000000    2.00000  
50%      1.000000    3.00000  
75%      2.000000    4.00000  
max      3.000000    5.00000

Столбец 4 (ректальня температура):
1. средняя температура - 38,1 
2. минимальная температура - 35,4
3. максимальная температура - 40,8
4. среднеквадратичное отклонение - 0,7
5. медиана - 38,2

Столбец 5 (пульс):
1. средний пульс - 71,9 
2. минимальный пульс - 30
3. максимальный пульс - 184
4. среднеквадратичное отклонение - 28,6
5. медиана - 64

Столбец 6 (частота дыхания)
1. Средняя частота - 30 вдохов
2. Минимальная частота - 8
3. Максимальная частота - 96
4. среднеквадратичное отклонение - 17,6
5. медиана - 24,5

Остальные столбцы содержат качественные показатели, в связи с чем данные метрики значимой информации не содержат

In [73]:
def moda(df):
    rec_temp = {}
    for w in df.round():
        if w not in rec_temp:
            rec_temp[w] = 1
        else:
            rec_temp[w] += 1

    maxt= 0
    mode_temp = None
    for k, v in rec_temp.items():
        if maxt < v:
            maxt = v
            mode_temp = k
    return f'Наибольшее количество раз ({maxt}) встречается показатель {mode_temp}'

#### Мода для столбца 4 (ректальня температура)

In [15]:
moda(df1[4])


'Наибольшее количество раз (152) встречается показатель 38.0'

#### Мода для столбца 5 (пульс)

In [16]:
moda(df1[5])


'Наибольшее количество раз (28) встречается показатель 48.0'

#### Мода для столбца 6 (частота дыхания)

In [17]:
moda(df1[6])


'Наибольшее количество раз (28) встречается показатель 20.0'

#### Мода для столбца 7 (температура конечностей)
При этом: 1 = Нормальная, 2 = Теплая, 3 = Прохладная,  4 = Холодная

In [25]:
moda(df1[7])


'Наибольшее количество раз (109) встречается показатель 3.0'

#### Мода для столбца 8 (пульс)
При этом:   1 = нормальный, 2 = повышенный, 3 = сниженный, 4 = отсутствует

In [27]:
moda(df1[8]) 

'Наибольшее количество раз (115) встречается показатель 1.0'

#### Мода для столбца 9 (состояние слизистых оболочек)
При этом:   1 = нормальный розовый, 2 = ярко-розовый, 3 = бледно-розовый, 4 = бледно-синюшный, 5 = ярко-красный / впрыснутый, 6 = темно-синюшный

In [28]:
moda(df1[9]) 

'Наибольшее количество раз (79) встречается показатель 1.0'

#### Мода для столбца 10  (время наполнения капилляров)
При этом:   1 = <3 секунд,
            2 = => 3 секунд

In [29]:
moda(df1[10]) 

'Наибольшее количество раз (188) встречается показатель 1.0'

#### Мода для столбца 11  (боль)
При этом: 1 = тревога, без боли, 2 = депрессия, 3 = прерывистая легкая боль, 4 = периодическая сильная боль, 5 = постоянная сильная боль

In [30]:
moda(df1[11]) 

'Наибольшее количество раз (67) встречается показатель 3.0'

### Задание 2. Работа с выбросами

In [19]:
df2 = horse.iloc[:,3:6]

In [35]:
def outliers(df):
    q1 = df.quantile(0.25)
    q3 = df.quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    remove_outliers = df2[df.between(lower_bound, upper_bound, inclusive=True)]
    return pd.concat([df2, remove_outliers]).drop_duplicates(keep=False)

In [41]:
outliers(df2[4]).sort_values(4).head(13)

4      5     6
44   35.4  140.0  24.0
141  36.0   42.0  30.0
238  36.1   88.0   NaN
80   36.4   98.0  35.0
118  36.5   78.0  30.0
298  36.5  100.0  24.0
251  36.6   42.0  18.0
99   39.6  108.0  51.0
75   39.7  100.0   NaN
20   39.9   72.0  60.0
281  40.0   78.0   NaN
259  40.8   72.0  42.0
7     NaN   60.0   NaN

При нормальной ректальной температуре лощади в 37,8 градусов, минимальная температура составляет 35,4 градуса, максимальная - 40,8 градуса, что соответсвует нормальным значениям, в связи с чем, выбросов по температуре не установлено.

In [42]:
outliers(df2[5]).sort_values(5)

4      5     6
41   39.0  150.0  72.0
275  38.8  150.0  50.0
55   38.6  160.0  20.0
3    39.1  164.0  84.0
255  38.8  184.0  84.0
58    NaN    NaN  20.0
115   NaN    NaN  40.0
117  39.5    NaN   NaN
126  38.7    NaN   NaN
151  39.3    NaN   NaN
159   NaN    NaN  12.0
204  39.2    NaN  23.0

При нормальном пульсе лощади в 30-40 ударов в минуту, пульс от 150 до 184 является явно значительно превышающим стандартный показатель, в связи с чем подлежит исключению из выборки либо изменению методом аналогичным при работе с пропусками.


In [46]:
outliers(df2[6]).sort_values(6).head(16)

4      5     6
120  39.4   54.0  66.0
103  38.0  140.0  68.0
125  38.0   42.0  68.0
265   NaN   56.0  68.0
229  38.5  120.0  70.0
295   NaN  120.0  70.0
41   39.0  150.0  72.0
82   38.1  100.0  80.0
84   37.8   60.0  80.0
208  37.8   88.0  80.0
3    39.1  164.0  84.0
255  38.8  184.0  84.0
244  38.2  124.0  88.0
39   39.2  146.0  96.0
106  38.3   52.0  96.0
7     NaN   60.0   NaN

В соответстии с описанием столбцов в наборе анализируемых данных, нормальная частота дыхания составляет от 8 до 10 вдохов, вместе с тем, также указано на низкую полезность данных показателей из-за их больших колебаний. При сопоставлении с ранее рассчитанными метриками, нормальная частота дыхания выявлена менее чем у 25 % лошадей, средняя частота дыхания - 30 вдохов, а мода - 20, в связи с чем выявленные выбросы подлежат включению в выборку для дальнейшего анализа и сопоставления с иными показателями.

### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [208]:
df3 = df1.rename({4:'temp', 5:'pulse',6:'respiratory',7:'temp_ext',8:'per_pulse',9:'membranes',10:'refill',11:'pain'}, axis=1)

In [190]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   temp         240 non-null    float64
 1   pulse        276 non-null    float64
 2   respiratory  242 non-null    float64
 3   temp_ext     244 non-null    float64
 4   per_pulse    231 non-null    float64
 5   membranes    253 non-null    float64
 6   refill       268 non-null    float64
 7   pain         245 non-null    float64
dtypes: float64(8)
memory usage: 18.9 KB


In [209]:
df3_drop = df3.dropna(thresh=5)

In [192]:
df3_drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 265 entries, 0 to 298
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   temp         220 non-null    float64
 1   pulse        251 non-null    float64
 2   respiratory  221 non-null    float64
 3   temp_ext     242 non-null    float64
 4   per_pulse    229 non-null    float64
 5   membranes    250 non-null    float64
 6   refill       258 non-null    float64
 7   pain         240 non-null    float64
dtypes: float64(8)
memory usage: 18.6 KB


In [210]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 17.0%
pulse - 5.3%
respiratory - 16.6%
temp_ext - 8.7%
per_pulse - 13.6%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


Исходя из описания таблицы можно усмотреть зависимость между столбцом 4  (ректальная температура - temp) и столбцом 7 (температура конечностей - temp_ext), в связи с чем возможно заполнить пропуски в столбце 4 значенииями медиан ректальной темпертуры по сгруппированному столбцу 7, а затем аналогичным образом заполнить столбец 7, сгруппировав столбец 4

In [211]:
df3_drop['temp'].fillna(df3_drop.groupby(['temp_ext']).temp.transform('median'), inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [212]:
df3_drop['temp_ext'].fillna(df3_drop.groupby(['temp']).temp_ext.transform('median'), inplace=True)

In [213]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.4%
pulse - 5.3%
respiratory - 16.6%
temp_ext - 0.8%
per_pulse - 13.6%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


In [214]:
df3_drop['temp'].fillna(df3_drop['temp'].mean(), inplace=True)

In [215]:
df3_drop['temp_ext'].fillna(df3_drop['temp_ext'].median(), inplace=True)

In [216]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 5.3%
respiratory - 16.6%
temp_ext - 0.0%
per_pulse - 13.6%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


Исходя из описания таблицы можно усмотреть зависимость между столбцом 5 (пульс- pulse) и столбцом 8 (периферический пульс  - per_pulse), в связи с чем возможно заполнить пропуски в столбце 5 значенииями медиан пульса по сгруппированному столбцу 8, а затем аналогичным образом заполнить столбец 8, сгруппировав столбец 5

In [217]:
df3_drop['pulse'].fillna(df3_drop.groupby(['per_pulse']).pulse.transform('median'), inplace=True)

In [218]:
df3_drop['per_pulse'].fillna(df3_drop.groupby(['pulse']).per_pulse.transform('median'), inplace=True)

In [219]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 0.0%
respiratory - 16.6%
temp_ext - 0.0%
per_pulse - 1.1%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


In [220]:
df3_drop['per_pulse'].fillna(df3_drop['per_pulse'].median(), inplace=True)

In [221]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 0.0%
respiratory - 16.6%
temp_ext - 0.0%
per_pulse - 0.0%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


Также можно определить прямую зависимость частоты дыхания от пульса, в связи с чем заполнить столбец 6 (частота дыхания - respiratory) способом аналогичным прыдущим

In [223]:
df3_drop['respiratory'].fillna(df3_drop.groupby(['pulse']).respiratory.transform('median'), inplace=True)

In [224]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 0.0%
respiratory - 0.8%
temp_ext - 0.0%
per_pulse - 0.0%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


In [226]:
df3_drop['respiratory'].fillna(df3_drop['respiratory'].median(), inplace=True)

In [227]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 0.0%
respiratory - 0.0%
temp_ext - 0.0%
per_pulse - 0.0%
membranes - 5.7%
refill - 2.6%
pain - 9.4%


Исходя из описания таблицы усматривается зависимость боли (столбец 11 - pain) от пульса (столбец 5 - pulse) и температуры конечностей (столбец 7 - temp_ext), в связи с чем можно заполнить попуски в столбце 11 медианами сгрупированных значений столбцов 5 и 7.


In [238]:
df3_drop['pain'].fillna(df3_drop.groupby(['pulse','temp_ext']).pain.transform('median'), inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [239]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 0.0%
respiratory - 0.0%
temp_ext - 0.0%
per_pulse - 0.0%
membranes - 5.7%
refill - 2.6%
pain - 1.9%


Оставшиеся пропуски, в связи с их небольшим соотношением к общему количеству, можно заполнить значениями медиан по соответсвующим столбцам.

In [240]:
df3_drop['membranes'].fillna(df3_drop['membranes'].median(), inplace=True)
df3_drop['refill'].fillna(df3_drop['refill'].median(), inplace=True)
df3_drop['pain'].fillna(df3_drop['pain'].median(), inplace=True)

In [241]:
print(f'Количество пропусков в столбцах:')
for col in df3_drop.columns:
    pct_missing = df3_drop[col].isna().mean()
    print(f'{col} - {pct_missing :.1%}')

Количество пропусков в столбцах:
temp - 0.0%
pulse - 0.0%
respiratory - 0.0%
temp_ext - 0.0%
per_pulse - 0.0%
membranes - 0.0%
refill - 0.0%
pain - 0.0%
